src/index.ts

```diff
@@ -0,0 +1,191 @@
+  | 'Amazon OS'
+  | 'Windows 3.11'
+  | 'Windows 95'
+  | 'Windows 98'
+  | 'Windows 2000'
+  | 'Windows XP'
+  | 'Windows Server 2003'
+  | 'Windows Vista'
+  | 'Windows 7'
+  | 'Windows 8'
+  | 'Windows 8.1'
+  | 'Windows 10'
+  | 'Windows ME'
+  | 'Open BSD'
+  | 'Sun OS'
+  | 'Linux'
+  | 'Mac OS'
+  | 'QNX'
+  | 'BeOS'
+  | 'OS/2'
+  | 'Search Bot';
+type UserAgentRule = [Browser, RegExp];
+type UserAgentMatch = [Browser, RegExpExecArray] | false;
+type OperatingSystemRule = [OperatingSystem, RegExp];
+
+const SEARCHBOX_UA_REGEX = /alexa|bot|crawl(er|ing)|facebookexternalhit|feedburner|google web preview|nagios|postrank|pingdom|slurp|spider|yahoo!|yandex/;
+const SEARCHBOT_OS_REGEX = /(nuhk)|(Googlebot)|(Yammybot)|(Openbot)|(Slurp)|(MSNBot)|(Ask Jeeves\/Teoma)|(ia_archiver)/;
+const REQUIRED_VERSION_PARTS = 3;
+
+const userAgentRules: UserAgentRule[] = [
+  ['aol', /AOLShield\/([0-9\._]+)/],
+  ['edge', /Edge\/([0-9\._]+)/],
+  ['yandexbrowser', /YaBrowser\/([0-9\._]+)/],
+  ['vivaldi', /Vivaldi\/([0-9\.]+)/],
+  ['kakaotalk', /KAKAOTALK\s([0-9\.]+)/],
+  ['samsung', /SamsungBrowser\/([0-9\.]+)/],
+  ['chrome', /(?!Chrom.*OPR)Chrom(?:e|ium)\/([0-9\.]+)(:?\s|$)/],
+  ['phantomjs', /PhantomJS\/([0-9\.]+)(:?\s|$)/],
+  ['crios', /CriOS\/([0-9\.]+)(:?\s|$)/],
+  ['firefox', /Firefox\/([0-9\.]+)(?:\s|$)/],
+  ['fxios', /FxiOS\/([0-9\.]+)/],
+  ['opera', /Opera\/([0-9\.]+)(?:\s|$)/],
+  ['opera', /OPR\/([0-9\.]+)(:?\s|$)$/],
+  ['ie', /Trident\/7\.0.*rv\:([0-9\.]+).*\).*Gecko$/],
+  ['ie', /MSIE\s([0-9\.]+);.*Trident\/[4-7].0/],
+  ['ie', /MSIE\s(7\.0)/],
+  ['bb10', /BB10;\sTouch.*Version\/([0-9\.]+)/],
+  ['android', /Android\s([0-9\.]+)/],
+  ['ios', /Version\/([0-9\._]+).*Mobile.*Safari.*/],
+  ['safari', /Version\/([0-9\._]+).*Safari/],
+  ['facebook', /FBAV\/([0-9\.]+)/],
+  ['instagram', /Instagram\s([0-9\.]+)/],
+  ['ios-webview', /AppleWebKit\/([0-9\.]+).*Mobile/],
+  ['searchbot', SEARCHBOX_UA_REGEX]
+];
+const operatingSystemRules: OperatingSystemRule[] = [
+  ['iOS', /iP(hone|od|ad)/],
+  ['Android OS', /Android/],
+  ['BlackBerry OS', /BlackBerry|BB10/],
+  ['Windows Mobile', /IEMobile/],
+  ['Amazon OS', /Kindle/],
+  ['Windows 3.11', /Win16/],
+  ['Windows 95', /(Windows 95)|(Win95)|(Windows_95)/],
+  ['Windows 98', /(Windows 98)|(Win98)/],
+  ['Windows 2000', /(Windows NT 5.0)|(Windows 2000)/],
+  ['Windows XP', /(Windows NT 5.1)|(Windows XP)/],
+  ['Windows Server 2003', /(Windows NT 5.2)/],
+  ['Windows Vista', /(Windows NT 6.0)/],
+  ['Windows 7', /(Windows NT 6.1)/],
+  ['Windows 8', /(Windows NT 6.2)/],
+  ['Windows 8.1', /(Windows NT 6.3)/],
+  ['Windows 10', /(Windows NT 10.0)/],
+  ['Windows ME', /Windows ME/],
+  ['Open BSD', /OpenBSD/],
+  ['Sun OS', /SunOS/],
+  ['Linux', /(Linux)|(X11)/],
+  ['Mac OS', /(Mac_PowerPC)|(Macintosh)/],
+  ['QNX', /QNX/],
+  ['BeOS', /BeOS/],
+  ['OS/2', /OS\/2/],
+  ['Search Bot', SEARCHBOT_OS_REGEX]
+];
+
+export function detect(): BrowserInfo | BotBrowserInfo | NodeInfo | null {
```

Comment:
Union types are helpful when we create type definitions downstream of libraries, but I don't think they're the best solution here now that you're baking typescript right into the library (thanks for that, by the way).

Consider this approach, which uses a single interface with three different implementations instead of three distinct interfaces. It lets us care less about what type is returned from `detect()` and instead focus on the values of the properties. It also guarantees all of various types conform to the same property names because they implement the same interface. Right now they align merely by convention.

You can still use these explicit implementations in `parseUserAgent` and `getNodeVersion` as appropriate. Here's an example. You might want to go one further and combine the `BrowserInfo` and `BotInfo` classes into a single implementation. 🤷‍♂️ 

```ts
// IProcessInfo?
interface IUserInfo {
  name: Browser | 'node' | null;
  version: string;
  os: OperatingSystem | NodeJS.Platform | null;
  bot: boolean;
}

class BrowserInfo implements IUserInfo {

  public bot: boolean = false;

  constructor(
    public name: Browser,
    public version: string,
    public os: OperatingSystem) {}

}

class NodeInfo implements IUserInfo {

  public name: 'node' = 'node';
  public os: NodeJS.Platform = process.platform;
  public bot: boolean = false;

  constructor(public version: string) {}

}

class BotInfo implements IUserInfo {

  public name: null = null;
  public version: string = null;
  public os: null = null;

  constructor(public bot: boolean) {}
  
}

export function detect(): IUserInfo | null { /* ... */ }
export function parseUserAgent(ua: string): BrowserInfo | BotInfo | null { /* ... */ }
export function getNodeVersion(): NodeInfo | null { /* ... */ }
```

Reply:
Good point re union types, though to be honest I'm kind of a fan. 

One thing I notice though now I'm reviewing my own code with somewhat fresh eyes is that the lack of a discriminator on the interfaces is going to make type-safe handling the result from detect function more onerous than I'd like.

I think at the very least I need to move to something like this:

```ts
interface BaseInfo {
  type: 'browser' | 'node' | 'bot';
}

interface BrowserInfo extends BaseInfo {
  type: 'browser';
  name: Browser;
  version: string;
  os: OperatingSystem | null;
}

interface NodeInfo extends BaseInfo {
  type: 'node';
  version: string;
  os: NodeJS.Platform;
};

interface BotInfo extends BaseInfo {
  type: 'bot';
}
```

Reply:
...actually I'd write it like this (coffee kicking in):

```ts
interface BaseInfo<T extends 'browser' | 'node' | 'bot'> {
  type: T;
}

interface BrowserInfo extends BaseInfo<'browser'> {
  name: Browser;
  version: string;
  os: OperatingSystem | null;
}

interface NodeInfo extends BaseInfo<'node'> {
  version: string;
  os: NodeJS.Platform;
};

interface BotInfo extends BaseInfo<'bot'> {
}
```

Reply:
This is moving types from three different classes of objects to a string-based enum named after a reserved keyword. IMHO, the more idiomatic way to check what type something is would be using the `instanceof` versus a property.

Also, given the above generic type definition, what would be the method signature of `detect()` with this specification? If you're suggesting this:

```ts
export function detect(): BrowserInfo | NodeInfo | BotInfo | null { /* ... */ }
```

I urge you to try writing some code consuming it. You're not going to be able to check for properties that only exist on one of the unioned types until you do a type check. That seems like an undesirable user experience.

```ts
let result = detect();

// The following line gives this error:
// Property 'os' does not exist on type 'BrowserInfo | NodeInfo | BotInfo'.
// Property 'os' does not exist on type 'BotInfo'. 
console.log(result.os);

if (result.type === 'browser') {
    console.log(result.os);  // but this works
}
```

Reply:
Agreed `type` might not be the best choice for the identifier.  If the people's vote is for classes then I'm happy to go that way... though I feel like it's possible overcooked for what is essentially a record of information.  Also not sure on the transpiled bundle size if classes get involved.

I agree that people coming from a strong OOP background would definitely reach for `instanceof` but I'm rarely doing that kind of thing at runtime if I have alternative strategies available.  Union types with a discriminator are an awesome way to go in this regard.

> I urge you to try writing some code consuming it. 

Yep - absolutely.  I guess my goal was to ensure very strong type safety from the module, but agree that does possibly make it easier to use from a "I just want to use this thing" approach.

... so currently doing that :)

Reply:
OK, here's what I think I'd really like to go with:
_(look away if you don't like complicated template / generic types - they certainly aren't as readable as classes)_

```ts
interface BaseInfo<N extends string, O = null, V = null> {
  name: N;
  os: O;
  version: V;
}

interface BrowserInfo extends BaseInfo<Browser, OperatingSystem | null, string> {
}

interface NodeInfo extends BaseInfo<'node', NodeJS.Platform, string> {
}

interface BotInfo extends BaseInfo<'bot'> {
}
```

I've tested using these interface definitions in some actual code and the intellisense behaviour is **awesome** (to be fair, it was probably just as good with your originally suggested classes).  Additionally, the behaviour with strict null checking (which I think everyone should have turned on) works as I'd hoped.  Horribly contrived example below:

```ts
import { detect } from 'detect-browser';

const result = detect();

if (result) {
  if (result.name === 'node') {
    console.log(result.os.length); // ok
  }

  console.log(result.os.length); // fail
  console.log(result.version.length); // fail
}
```

I'm going to commit there changes off to a separate branch, and do a spike on the class implementation also....

Reply:
@carusology Thanks heaps for the review mate.  I think the class based implementation (with a bit of template typing build in for good measure is likely the winner - see #81).

Key points:

1. The size of the bundle is not dramatically different, i.e. less that 100 bytes (i.e. nothing) more:
    - template types only (1877 bytes gzipped)
    - classes (1964 bytes gzipped)
2. Folks have the ability to use the `instanceof` operator if they wish, which means they get some nice choice when consuming the module.

If you wouldn't mind having a look at #81 and letting me know your thoughts.  Assuming you like the changes, I'll merge that #81 into this PR and we can go from there.  I think with your feedback it's pretty close to being ready for primetime.

Reply:
Oh, and intellisense behaviour is just as good.

Reply:
Some 🥃influenced feedback:

1. #80 and #81 are not _quite_ comparing child interfaces versus class implementations in isolation. #80 does not export the child interfaces, while #81 exports the classes. In an ideal world, these child implementations should not be necessary for export since the abstraction should be sufficient, but they may also have use to your downstream consumers. I think this is an appropriate place to default to your opinions since you own the module.

2. I am not pushing for classes per se. The thing I am advocating is that consumers should not need to know which of the union types they are dealing with in order to consume the results of the `detect()` method. For this reason, I would advocate for the `bot` property to move up to `BaseInfo` (in #80) or `DetectedInfo` (in #81) as a `boolean` type, with `BotInfo`, be it a class or interface, returning `true`, and the other child classes (or interfaces) returning `false`. Right now consumers still have to know what type they're dealing with to check the `bot` property:

```ts
let result = detect();

// TS2339: Property 'bot' does not exist on type 'BrowserInfo | NodeInfo | BotInfo'.
// Property 'bot' does not exist on type 'BrowserInfo'.
console.log(result.bot);  // fails

// I don't think I should have to do this check first.
// It makes the subsequent check superfluous.
if (result.name === 'bot') {
  console.log(result.bot); // works
}
```

Teeing off #81, I'm advocating something like this:

```ts
interface DetectedInfo<N extends string, O, V = null> {
  name: N;
  version: V;
  os: O;
  bot: boolean;
}

export class BrowserInfo implements DetectedInfo<Browser, OperatingSystem | null, string> {
  constructor(
    public readonly name: Browser,
    public readonly version: string,
    public readonly os: OperatingSystem | null,
    public readonly bot: false = false) {
  }
}

export class NodeInfo implements DetectedInfo<'node', NodeJS.Platform, string> {
  public readonly name: 'node' = 'node';
  public readonly os: NodeJS.Platform = process.platform;
  public readonly bot: false = false;

  constructor(public readonly version: string) {
  }
}

export class BotInfo implements DetectedInfo<'bot', null, null> {
  public readonly bot: true = true;
  public readonly name: 'bot' = 'bot';
  public readonly version: null = null;
  public readonly os: null = null;
}

let result = detect();
console.log(result.bot);  // NOW this works
```

3. It looks like you want the properties on the objects that are returned to be `readonly`. If desired, you can set that on the interface to require it.

```ts
interface DetectedInfo<N extends string, O, V = null> {
  readonly name: N;
  readonly version: V;
  readonly os: O;
  readonly bot: boolean; // If you take my second suggestion
}
```

4. Years of C# development have taught me to name my generics as `TFoo`, as opposed to single characters, to make it more obvious what they refer to. It would be something to consider here to address the readability concerns you highlighted in your comment above:

```ts
interface IDetectedInfo<TName extends string, TOperatingSystem, TVersion = null> {
  readonly name: TName;
  readonly version: TVersion;
  readonly os: TOperatingSystem;
}
```

Reply:
OK, getting by head back into this after a bit of a break over Christmas and New Year.  Going to keep this pretty short as I think we've had some good discussion and I've changed a few things as a result.

- Definitely going with the implementation using classes.  It's a good approach which gives people a bit more flexibility in how they use the module.
- Not going to promote the `bot` flag to the interface.  I'm going to mark this as deprecated in the next major release (tempted to remove it entirely) because there really is 2 ways now to check for a bot with the core constructs, having a third for sugar seems like overkill.
- Using `readonly` on the interface (that was a fantastic suggestion).
- Probably not going to mark the template placeholders with `TFoo`.  Mainly personal preference for keeping things consistent (even if more confusing) with the current Typescript ecosystem... at least the parts that I've seen.

Will be squash merging #81 and publishing a new major release in the next day or two as I'm keen to :shipit:

Reply:
@carusology suggestion to try examples with real code is key:

> I urge you to try writing some code consuming it. 

And the idea to roll with classes over interfaces despite a small amount of bloat to the final bundle is warranted given the flexibility it provides consumers of the lib.

PR #81 looks solid. Nice work.

Reply:
I think it's fine to ship as is, but I do have a question on this:

>Not going to promote the `bot` flag to the interface. I'm going to mark this as deprecated in the next major release (tempted to remove it entirely) because there are 2 ways now to check for a bot with the core constructs, having a third for sugar seems like overkill.

Is this for JavaScript consumers? As shown in my previous code snippet, the consumer has to know it's an instance of `BotInfo` before he or she can check the `bot` property, which makes the property's continued existence superfluous in TypeScript settings. However, in JavaScript, I acknowledge these types will not exist, so the consumers may not expect this breaking change even if the version change is a major one.

If you want to warn users that this is deprecated in a JavaScript setting where they lack these types, you could make the `bot` property [a get accessor](https://www.typescriptlang.org/docs/handbook/classes.html) that spits out a warning to the console about its impending removal in the next major release. [You're targeting ECMAScript5](https://github.com/DamonOehlman/detect-browser/pull/81/files#diff-e5e546dd2eb0351f813d63d1b39dbc48R3), so this is supported with your current compilation options.

Here's an example:
```ts
export class BotInfo implements DetectedInfo<'bot', null, null> {
   private _isWarningEmitted: boolean = false;

   public get bot(): boolean {
       if (!this._isWarningEmitted) {
           this._isWarningEmitted = true;
           console.warn('The "bot" property is deprecated. It will be removed in the next major release.');
       }

       return true;
   }

   public readonly name: 'bot' = 'bot';
   public readonly version: null = null;
   public readonly os: null = null;
 }
```

Anyway, I think the UX around this composition is fine in a TypeScript setting as TypeScript users will be sheparded against using the deprecated `bot` property by how you have structured the properties. However, if you want to wean JavaScript users off this property gradually, the above suggestion is something to consider.

Reply:
Another good point... and yes, I expect that the primary consumers of the `detect-browser` package will be JS users.  That said, the falsiness of JS should make for a graceful transition (for most people).

The absence of the `bot` flag on those non-bot records _I think_ be handled correctly for all existing use cases.  The only corner case is if something is actually doing a `result.bot == false` check (or `===` check), and I'd be pretty surprised by the existence of that in JS codebases. 

Even so, it is a breaking change, and thus the major version will be bumped and this adequately documented in the release notes.  I think I'll specifically call out the `result.bot == false` (and triple equals case alse) as the only case that should fail.  I can appreciate that including it across all the result types would be such a small thing, but I really do want to move away from having boolean flags exposed like this moving forward.  It was a mistake on my part to include it in the first place.

Reply:
Draft release notes:

https://github.com/DamonOehlman/detect-browser/releases/untagged-28fdb805d384f757a166#

/cc @sargant